pkg/evaluator/ranger.go

```diff
@@ -0,0 +1,78 @@
+package evaluator
+
+type ranger interface {
+	next() bool
+}
+
+type stepRange struct {
+	loopVar *Num
+	cur     float64
+	stop    float64
+	step    float64
+}
+
+type arrayRange struct {
+	loopVar Value
+	cur     int
+	array   *Array
+}
+
+type mapRange struct {
+	loopVar Value
+	cur     int // index of Map.Order slice of keys
+	mapVal  *Map
+	order   []string // copy of order in case map entry gets deleted during iteration
+}
+
+type stringRange struct {
+	loopVar *String
+	cur     int
+	str     *String
+	runes   []rune
+}
+
+func (s *stepRange) next() bool {
+	if s.step > 0 && s.cur >= s.stop {
+		return false
+	}
+	if s.step < 0 && s.cur <= s.stop {
+		return false
+	}
+	s.loopVar.Val = s.cur
+	s.cur += s.step
+	return true
+}
+
+func (a *arrayRange) next() bool {
+	elements := *a.array.Elements
+	if a.cur >= len(elements) {
+		return false
+	}
+	a.loopVar.Set(elements[a.cur])
+	a.cur += 1
```

Comment:
Could be `a.cur++`. I sometimes wonder why Go has `++` - may as well do this `+= 1` since you can't use `++` in an expression. Up to you.

Reply:
ok, updated.